Set :host width as well

tidoust opened the pull request:
As a block element, the `<i-slide>` was taking the entire width available, even though its content width is clamped to the value of the `width` attribute (or 300px). With this update, the width of the element is explicitly set to the width of its content. That is more consistent with the way `<img>` and `<iframe>` work.

dontcallmedom commented:
> That is more consistent with the way `<img>` and `<iframe>` work.

but both  `<img>` and `<iframe>`  defaults to being `inline`-displayed elements, so I'm not sure the comparison holds.

tidoust commented:
I was under the wrong assumption that `<iframe>` was a block element, but that was only the rule for the old `<frame>` and `<frameset>` in HTML4.

This takes me back to the question that I started to wonder about in the past couple of days: shouldn't `<i-slide>` be an inline element instead of a block element? I thought block was the answer to sort of mimic `<iframe>`, but since that's actually not true for `<iframe>`, why use block at all?

Asked differently: what do we need to consider to decide whether a custom element should be an inline or block element by default?

Malvoz commented:
We've been trying to mimic iframe/video with `<mapml-viewer>` ([example](https://maps4html.org/MapML/spec/#example-1)), [we're using `inline-block`](https://github.com/Maps4HTML/Web-Map-Custom-Element/blob/f3e4390f513fc7b5e79fec4f098925ba1bd07953/src/mapml-viewer.js#L120) as a compromise because we have to set a default size of 300x150 in CSS px but keep the inline behavior of the aforementioned elements, and `inline` can't have height or width.

dontcallmedom commented:
> Asked differently: what do we need to consider to decide whether a custom element should be an inline or block element by default?

I was wondering the same - I think it's part of the broader question of what CSS questions should a Web component answer, or at least those have aim at universality of usage (which I'm calling "browser-grade" Web components). I guess that we should look at what CSS properties get set by UA stylesheets à la [Firefox'](https://searchfox.org/mozilla-central/source/layout/style/res/html.css).

> Shouldn't `<i-slide>` be an inline element instead of a block element? I thought block was the answer to sort of mimic `<iframe>`, but since that's actually not true for `<iframe>`, why use block at all?

For `i-slide` specifically, maybe we should indeed make it an inline element (to mimic `img` and  `iframe`); `inline-block` is tempting too, but it seems only `marquee` (!) is set to this as a default.

tidoust commented:
I guess one drawback to inline for `<i-slide>` is the need to wrap it in a block element (or to add an `i-slide { display: block; }` directive in the containing document) in most cases as the usage scenarios that we've identified so far are more going to use the custom element as a block. Not sure if that's a good enough rationale to justify using block. Another one

I note some related past discussion on [whether custom elements should remain `display: inline` by default](https://github.com/WICG/webcomponents/issues/224). The discussion does not really conclude one way or the other.

@Malvoz Thanks for sharing. I haven't tried it yet but I'm wondering whether inline-block is actually required. That is, it should be possible to leave the custom element as an inline element and rather set the root element of the shadow tree as a block element with proper dimensions, to create what CSS refers to as a [replaced element](https://www.w3.org/TR/css-display/#replaced-element). Or maybe not?

tidoust commented:
> @Malvoz Thanks for sharing. I haven't tried it yet but I'm wondering whether inline-block is actually required. That is, it should be possible to leave the custom element as an inline element and rather set the root element of the shadow tree as a block element with proper dimensions, to create what CSS refers to as a [replaced element](https://www.w3.org/TR/css-display/#replaced-element). Or maybe not?

Well, maybe not :) I cannot find a way to properly create such a replaced element, in other words a custom element that appears as inline from the perspective of the containing document, but which still has intrinsic dimensions. I manage to get close to that if I leave the custom element as an inline element and set the root in the shadow tree to `display: inline-block` but the height of the `<i-slide>` element remains the height of a line of text, which seems wrong.

tidoust commented:
> Authors are likely going to want to use `block` most of the time, which is also true for iframe, video, img, etc., though authors are used to overriding the default of these elements. So while a default of `block` would be more convenient to authors, `inline-block` behaves more similarly to standard replaced elements. I decided on `inline-block` for the MapML-viewer component simply because I expect a standard map viewer would have the inline behavior.

I concur. Making the custom element `inline-block` gives the closest thing to a proper inline `<i-slide>` element which is, as for your MapML viewer element, what I would expect for it if it were a real HTML element.

@dontcallmedom What do you think?

dontcallmedom commented:
`inline-block` SGTM.

Out of curiosity, what do we think that kind of change implies for semversioning? (given we're brand new, I don't think we necessarily need to apply the conclusion we would reach here, I'm just trying to understand questions we may have to face in the future)

tidoust commented:
I think that would be a major bump from a sermversioning perspective because element will no longer generate line breaks before/after it, which means that HTML code that uses `<i-slide>` will likely need to change as a result (it may not if the element is surrounded by block elements, but we cannot assume that).

dontcallmedom commented:
Thinking some more about semver: the [rule is](https://semver.org/#introduction)
> For this system to work, you first need to declare a public API. This may consist of documentation or be enforced by the code itself. Regardless, it is important that this API be clear and precise.

There are two ways to think about how it applies to us:
* we define an API contract specific to `i-slide` (ideally baked in the test suite), and we bump major each time we change an existing rule of the contract
* we rely on the API contract that is (implicitly?) associated with being an HTML element (or in this case, a "browser-grade Web component")

Your suggestion this would be a major bump aligns with the latter, and furthermore implies that the default value of the CSS display property is part of that contract for HTML elements - both sound reasonable, but I would like us (not in this particular pull request :) to figure out if there is an explicit contract here, and if not, figure out how to make the implicit contract on which HTML elements have been designed both explicit and encoded in our test suite.

dontcallmedom commented:
some more rambling - technically, except in the very early history of HTML (when e.g. `<p>` switched from a being a self-closing element to be a containing element), I'm not sure HTML elements ever get the equivalent of a major bump - instead, new elements get created to provide the equivalent semantics but with the right model, e.g.  ̀applet`, `embed`, `object`, or `img` & `picture`. So if the rules of continuity aren't explicitly defined anywhere, these "renames" may be an historical source of understanding of the implicit rules.

dontcallmedom commented on i-slide.js:
```diff
@@ -235,6 +235,7 @@ class ISlide extends HTMLElement {
       return `
         :host {
           display: block;
```
```suggestion
          display: inline-block;
```

tidoust replied:
I'll do that but will also rebase and update the demo and the rest of the code, given that the switch means that there needs to be a wrapping block element somewhere.

tidoust commented:
I updated the PR accordingly to make `<i-slide>` an render as an inline block and updated tests, demo and README accordingly. I note that another consequence is that, when you wrap the element in a `<p>`, the `<p>` element is a few pixels taller than the `<i-slide>` element. That is normal, consistent with iframes and images, and due to default baseline adjustment rules. Setting `vertical-align: top` on the custom element aligns the planets (at least their heights).

tidoust commented:
@dontcallmedom I also note that the PR against scribe2 likely needs to be updated to add the wrapping block element (or a `display: block` rule): https://github.com/w3c/scribe2/pull/10

dontcallmedom commented:
shall we push a 1.1 tag?
